content/client-lib-development-guide/features.textile

```diff
@@ -299,7 +299,9 @@ h3(#realtime-connection). Connection
 ** @(RTN15b)@ In order for a connection to be resumed and connection state to be recovered, the client library reconnects to the websocket endpoint with two additional querystring params:
 *** @(RTN15b1)@ @resume@ is the private connection key assigned to the connection when the first @CONNECTED@ @ProtocolMessage@ was received
 *** @(RTN15b2)@ @connection_serial@ is the most recent @ProtocolMessage#connectionSerial@ received from Ably or @Connection#serial@ which should be identical
-** @(RTN15c)@ When a new connection is made using the @resume@ and @connection_serial@ params, the client library must consider the resume to have failed if the @CONNECTED@ @ProtocolMessage@ contains an error or if the @Connection#id@ has changed. When a @Connection@ resume has failed, all channels should be detach with a suitable error reason
+** @(RTN15c)@ The system's response to a resume request can be a @CONNECTED@ message with an error. If the @CONNECTED@ message @connectionId@ matches the prior @connectionId@, then Ably has confirmed to the client that the resume has been successful, the connection state is intact, and all attached channels remain attached. However, if the @connectionId@ has changed and optionally an error is present, then there may have been a non-fatal error (such as the system is unable to send the backlog in full because the limit of backlog messages has been exceeded), at which point the client library must assume that all connection state is lost and a new connection has been initialised
```

Comment:
> We are deterministic and predictable within the capacity limits of the infrastructure. If the thing that the developer asked to do exceeds the limits of the infrastructure, then we indicate that.

Well do we, because the spec does not even specify which error code represents which problem, and thus what action the developer will take.

> We will send the most recent 10,000 (or whatever) messages in the backlog; not a random sample. 

And why are the 10k most recent more valuable than the previous ones.  Can we really assume that is what the developer wants?

>  That IMO is a much more useful default behaviour than killing the connection and expecting the developer to pick up the pieces

I would agree that it's more useful in a lot of cases, but not in all cases. Also, if the buffer limit is hit, we can arguably save considerable resources by simply discarding that connection state.  I would prefer to tell users that we retain connection state for 120s and for up to 10k messages, and if either of those thresholds are exceeded they will be responsible for resuming the connection state or using history if the messages are important to them.  We have also recently discussed a system that will guarantee connection state recovery on channels that are persisted.  I feel that is a much better solution.

> Query each attached channel separately(does he know which serial to query from?) and reassemble those messages in connection order; 

As above, this is a planned feature for persisted channels.

> and that will possibly also fail because the limitations of his connection prevent him from retrieving a huge list in a useful timescale.

I am not sure I understand why that is.  If the message history is so important a developer will either need a guarantee that none are lost and he/she can retrieve them, or not care that much if some are lost and therefore losing connection state is not a train smash and they can continue.

I really dislike the idea that our behaviour is unpredictable. Perhaps we could offer an additional option that states if you hit the message limit for a connection we will discard the old messages, but I don't think that should be the default behaviour.